// edge.js/build/src/Edge/globals/index.js
"use strict";
/*
 * edge.js
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBALS = void 0;
const helpers_1 = require("@poppinss/utils/build/helpers");
const edge_error_1 = require("edge-error");
const Template_1 = require("../../Template");
exports.GLOBALS = {
    inspect: (value) => {
        return (0, Template_1.safeValue)(require('@poppinss/inspect').string.html(value));
    },
    truncate: (value, length = 20, options) => {
        options = options || {};
        return helpers_1.string.truncate(value, length, {
            completeWords: options.completeWords !== undefined ? options.completeWords : !options.strict,
            suffix: options.suffix,
        });
    },
    raise: (message, options) => {
        if (!options) {
            throw new Error(message);
        }
        else {
            throw new edge_error_1.EdgeError(message, 'E_RUNTIME_EXCEPTION', options);
        }
    },
    excerpt: (value, length = 20, options) => {
        options = options || {};
        return helpers_1.string.excerpt(value, length, {
            completeWords: options.completeWords !== undefined ? options.completeWords : !options.strict,
            suffix: options.suffix,
        });
    },
    safe: Template_1.safeValue,
    camelCase: helpers_1.string.camelCase,
    snakeCase: helpers_1.string.snakeCase,
    dashCase: helpers_1.string.dashCase,
    pascalCase: helpers_1.string.pascalCase,
    capitalCase: helpers_1.string.capitalCase,
    sentenceCase: helpers_1.string.sentenceCase,
    dotCase: helpers_1.string.dotCase,
    noCase: helpers_1.string.noCase,
    titleCase: helpers_1.string.titleCase,
    pluralize: helpers_1.string.pluralize,
    toSentence: helpers_1.string.toSentence,
    prettyBytes: helpers_1.string.prettyBytes,
    toBytes: helpers_1.string.toBytes,
    prettyMs: helpers_1.string.prettyMs,
    toMs: helpers_1.string.toMs,
    ordinalize: helpers_1.string.ordinalize,
};
